src/resolvers/Service.js

```diff
@@ -0,0 +1,207 @@
+ * - manager can only modify details
+ * - owner can create and delete + other operations
+ * @param {USER_TYPE} type
+ * @param {EMPLOYEE_TYPE} employeeType
+ */
+const roleValidator = (type, employeeType) => {
+  return (
+    type === USER_TYPE.BRANDOWNER ||
+    (type === USER_TYPE.EMPLOYEE && employeeType === EMPLOYEE_TYPE.MANAGER)
+  );
+};
+
+const estimatedServiceTimeConverter = (service) => {
+  service._doc.estimatedServiceTime = service._doc.estimatedServiceTime.toString();
+  return service;
+};
+
+const ServiceResolver = {
+  Query: {
+    services: async (root, { filter }, context, info) => {
+      /**
+       * - validate filter
+       * - try parse filter
+       * - retrieve
+       */
+
+      /**
+       * validate filter
+       */
+      if (filter === null) {
+        throw new UserInputError("No filter provided");
+      }
+
+      /**
+       * try parse filter
+       */
+      try {
+        filter = JSON.parse(filter);
+      } catch (error) {
+        throw new ApolloError("Unable to resolve filter", 500);
+      }
+
+      /**
+       * retrieve with filter
+       */
+      let services = await Service.find(filter);
+      services.map((service) => estimatedServiceTimeConverter(service));
+
+      return services;
+    },
+    service: async (root, { id }, context, info) => {
+      /**
+       * - validate, disallow if provided id
+       * - allow all access
+       */
+
+      /**
+       * must have ID
+       */
+      if (id === null) {
+        throw UserInputError("No ID provided to search");
+      }
+
+      /**
+       * validate for existing
+       */
+      let existingService = await Service.findById(id);
+      if (existingService === null) {
+        throw UserInputError("no service with this ID");
+      }
+      existingService = estimatedServiceTimeConverter(existingService);
+
+      return existingService;
+    },
+  },
+  Mutation: {
+    createService: async (root, { serviceInput }, context, info) => {
+      /**
+       * - validate, disallow if provided id
+       * - allow all access
+       * - validate for existing name
+       */
+
+      /**
+       * disallow ID
+       */
+      if (serviceInput.id) {
+        throw UserInputError("Received id at create mutatiion");
+      }
+
+      /**
+       * validate for existing name
+       */
+      const existingService = await Service.findOne({
+        serviceNm: { $eq: serviceInput.serviceNm },
+      });
+      // suggest to format the name string like brandID + serviceName to enable same name for different brands
+      if (existingService) {
+        throw UserInputError("Name already taken");
+      }
+
+      /**
+       * create if passes all validations
+       */
+      let newService = await Service.create(serviceInput);
+      newService = estimatedServiceTimeConverter(newService);
+
+      return newService;
+    },
+    updateService: async (root, { serviceInput }, context, info) => {
```

Comment:
validation on service name duplicated is not done here
suggest: create a const method for the service name validation check
something like roleValidator

Reply:
ok agree